### PR TITLE
Add CutMix method

### PR DESCRIPTION
**Pull request by coryMosaicML**
This adds CutMix as a method for vision classification problems 

- [x] CutMix implementation
- [x] Tests
- [x] Example runs: [ResNet50 Imagenet 77.25%](https://wandb.ai/mosaic-ml/cory-cutmix/runs/30zqnzvl), [ResNet56 CIFAR10 93.85%](https://wandb.ai/mosaic-ml/cory-cutmix/runs/175e0i9o)
- [x] Method card

### CODE REVIEW

**Review by growlix — COMMENTED** [review]
Implementation LGTM, just had some comments re: docs and clarity.

File: composer/algorithms/cutmix/cutmix.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    if cutmix_lambda is None:
+        cutmix_lambda = gen_cutmix_lambda(alpha)
+    if bbox:
+        rx, ry, rw, rh = bbox[0], bbox[1], bbox[2], bbox[3]
+    else:
+        rx, ry, rw, rh = rand_bbox(x.shape[2], x.shape[3], cutmix_lambda)
+        bbox = (rx, ry, rw, rh)
+
+    # Fill in the box with a part of a random image.
+    x_cutmix[:, :, rx:rw, ry:rh] = x_cutmix[shuffled_idx, :, rx:rw, ry:rh]
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply composer/algorithms/cutmix/cutmix.py:176]: I'm not sure anyone has experimented with doing it this way (see next comment).

File: docs/source/method_cards/cutmix.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+## Example Effects
+
+CutMix is intended to improve generalization performance, and we empirically find this to be the case in our image classification settings. The [original paper](https://arxiv.org/abs/1905.04899) also reports improvements in object localization and robustness.
+
+## Implementation Details
+
+The samples are created from a batch `(X, y)` of (inputs, targets) together with version `(X', y')` where the ordering of examples has been shuffled. The examples are combined by sampling a value `lambda` (between 0.0 and 1.0) from the Beta distribution parameterized by `alpha`, choosing a rectangular box within `X`, filling it with the data from the corresponding region in `X`. and training the network on the interpolation between `(X, y)` and `(X', y')`.
+
+Note that the same `lambda` and rectangular box are used for each example in the batch. Similar to MixUp, using the shuffled version of a batch to generate mixed samples allows CutMix to be used without loading additional data.
```
[the remark this answers is not in the document]
Comment (coryMosaicML) [reply docs/source/method_cards/cutmix.md:28]: In the code from the original implementation, I believe they also do it this way (same box for all examples in a batch). See here: https://github.com/clovaai/CutMix-PyTorch/blob/2d8eb68faff7fe4962776ad51d175c3b01a25734/train.py#L235

[the remark this answers is not in the document]
Comment (growlix) [reply docs/source/method_cards/cutmix.md:28]: Yup, looks like you're right! I misunderstood the description of the algorithm in the Methods section.

**Comment by growlix** [comment]
LGTM, Thanks for the clarification!

**Review by growlix — APPROVED** [review]

**Review by Landanjs — COMMENTED** [review]
I was to slow 🥲

Overall LGTM, there were two typos and a few nits that may not need to be fixed.

One thing is the method card needs to be added to index.rst (I think? @hanlint @growlix) to be viewed on readthedocs. It would be good to check how the method card looks on readthdocs before merging (welp, it's too late).

**Pull request merged**